### PR TITLE
[#12233] feat(storage): Store tag assignment values

### LINKED ISSUE #12233: [Subtask] Store tag assignment values
### Describe the subtask

Store tag assignment values in the relational storage layer.

### Parent issue

https://github.com/apache/gravitino/issues/12127

### PR DESCRIPTION
**Pull request by roryqi**
### What changes were proposed in this pull request?

This PR stores tag assignment values in the relational storage layer.

Changes include:
- Reuse `TagEntity` to expose assignment values in tag-object relation context.
- Persist tag value constraints in tag metadata storage.
- Persist assignment values in `tag_metadata_object_rel`.
- Support listing associated metadata objects by exact tag assignment value.
- Update H2/MySQL/PostgreSQL schema files and relational SQL providers.

### Why are the changes needed?

This is needed to support assigning multiple values to a tag and retrieving assigned tag values from storage.

Fix: #12233

### Does this PR introduce _any_ user-facing change?

No new public API is introduced in this PR. It adds relational storage support for tag value constraints and tag assignment values used by the tag value APIs.

### How was this patch tested?

- `./gradlew :core:spotlessApply`
- `./gradlew :core:test --tests org.apache.gravitino.meta.TestTagEntity --tests org.apache.gravitino.storage.relational.service.TestTagMetaService -PskipITs -PskipDockerTests=false`

### CODE REVIEW

File: scripts/mysql/schema-2.0.0-mysql.sql [hunk 1]
```diff
@@ -286,6 +286,7 @@ CREATE TABLE IF NOT EXISTS `tag_meta` (
     `metalake_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'metalake id',
     `tag_comment` VARCHAR(256) DEFAULT '' COMMENT 'tag comment',
     `properties` MEDIUMTEXT DEFAULT NULL COMMENT 'tag properties',
+    `allowed_values` MEDIUMTEXT DEFAULT NULL COMMENT 'tag allowed values',
```
Comment (yuqi1129) [inline scripts/mysql/schema-2.0.0-mysql.sql]: What's the meaning of allowed values here, and what's the value format for it.

Comment (roryqi) [reply scripts/mysql/schema-2.0.0-mysql.sql]: This value will be a string splitted by `,`.

Comment (roryqi) [reply scripts/mysql/schema-2.0.0-mysql.sql]: Tag allow values can be  a limitted collection. For example, if a tag named as `country`, the allow values can be `China,Japan,...`.

Comment (yuqi1129) [reply scripts/mysql/schema-2.0.0-mysql.sql]: OK, you may need to add information like the separator it uses.

Comment (yuqi1129) [reply scripts/mysql/schema-2.0.0-mysql.sql]: I notice that you are using jackson serialization, so the value format will be `["a","b","c"]`

File: core/src/main/java/org/apache/gravitino/storage/relational/utils/POConverters.java [hunk 2]
```diff
@@ -1465,8 +1472,34 @@ public static TagPO updateTagPOWithVersion(TagPO oldTagPO, TagEntity newEntity)
...
+    return allowedValues == null
+        ? null
+        : JsonUtils.anyFieldMapper().writeValueAsString(allowedValues);
+  }
+
+  private static String[] allowedValuesForStorage(TagValueConstraint valueConstraint) {
+    switch (valueConstraint.type()) {
+      case ANY_VALUE:
+        return null;
+      case NO_VALUE:
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/utils/POConverters.java:1488]: What's the difference between `ANY_VALUE` and `NO_VALUE`?

File: scripts/h2/schema-2.0.0-h2.sql [hunk 3]
```diff
@@ -308,14 +309,15 @@ CREATE TABLE IF NOT EXISTS `tag_relation_meta` (
     `tag_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'tag id',
     `metadata_object_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'metadata object id',
     `metadata_object_type` VARCHAR(64) NOT NULL COMMENT 'metadata object type',
+    `tag_value` VARCHAR(256) DEFAULT NULL COMMENT 'tag relation value',
```
Comment (yuqi1129) [inline scripts/h2/schema-2.0.0-h2.sql]: Do you need to modify the upgrade scripts too?

File: scripts/mysql/schema-2.0.0-mysql.sql [hunk 4]
```diff
@@ -299,14 +300,15 @@ CREATE TABLE IF NOT EXISTS `tag_relation_meta` (
     `tag_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'tag id',
     `metadata_object_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'metadata object id',
     `metadata_object_type` VARCHAR(64) NOT NULL COMMENT 'metadata object type',
+    `tag_value` VARCHAR(256) DEFAULT NULL COMMENT 'tag relation value',
     `audit_info` MEDIUMTEXT NOT NULL COMMENT 'tag relation audit info',
     `current_version` INT UNSIGNED NOT NULL DEFAULT 1 COMMENT 'tag relation current version',
     `last_version` INT UNSIGNED NOT NULL DEFAULT 1 COMMENT 'tag relation last version',
     `deleted_at` BIGINT(20) UNSIGNED NOT NULL DEFAULT 0 COMMENT 'tag relation deleted at',
     PRIMARY KEY (`id`),
-    UNIQUE KEY `uk_ti_mi_mo_del` (`tag_id`, `metadata_object_id`, `metadata_object_type`, `deleted_at`),
```
Comment (yuqi1129) [inline scripts/mysql/schema-2.0.0-mysql.sql:307]: Why do you remove the unique key?

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 5]
```diff
@@ -394,6 +514,99 @@ public int deleteTagMetasByLegacyTimeline(long legacyTimeline, int limit) {
...
+
+  private static void validateAllowedValue(TagPO tagPO, TagValue tagValue)
+      throws JsonProcessingException {
+    if (tagPO.getAllowedValues() == null) {
+      return;
+    }
+
+    String[] allowedValues =
+        JsonUtils.anyFieldMapper().readValue(tagPO.getAllowedValues(), String[].class);
+    if (!tagValue.value().isPresent()) {
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:596]: The value for `tagValue` can be `TagValue.noValue("stage")` and can pass the check when `TagValueConstraint.ofAllowedValues("dev", "prod")`

File: scripts/mysql/schema-2.0.0-mysql.sql (hunk 1 above)
Comment (roryqi) [reply scripts/mysql/schema-2.0.0-mysql.sql]: Fixed in 25b6f3fd2. The schema comments now state that `allowed_values` is stored as a JSON string array. Storage encoding is: `NULL` => `ANY_VALUE`, `[]` => `NO_VALUE`, and `["a","b"]` => allowed values.

File: scripts/mysql/schema-2.0.0-mysql.sql (hunk 4 above)
Comment (roryqi) [reply scripts/mysql/schema-2.0.0-mysql.sql:307]: The old unique key was on `(tag_id, metadata_object_id, metadata_object_type, deleted_at)`, so it would reject multiple active rows for the same tag-object assignment when different `tag_value`s are assigned. The service layer already de-duplicates requested active values, and the table keeps the primary key plus the lookup indexes.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java (hunk 5 above)
Comment (roryqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:596]: Fixed in 25b6f3fd2. `TagValue.noValue("stage")` is now rejected when the tag has a non-empty allowed-values constraint, and I added a regression test for that case in `TestTagMetaService`.

File: scripts/h2/schema-2.0.0-h2.sql (hunk 3 above)
Comment (roryqi) [reply scripts/h2/schema-2.0.0-h2.sql]: Fixed in 25b6f3fd2. Added 1.3.0 -> 2.0.0 upgrade changes for MySQL, H2, and PostgreSQL: add `tag_meta.allowed_values`, add `tag_relation_meta.tag_value`, add the `(tag_id, tag_value)` index, and drop the old tag-object unique constraint.

File: core/src/main/java/org/apache/gravitino/storage/relational/utils/POConverters.java (hunk 2 above)
Comment (roryqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/utils/POConverters.java:1488]: The difference is in the storage encoding and validation semantics: `ANY_VALUE` is stored as `NULL` and accepts valueless or any non-empty valued assignment; `NO_VALUE` is stored as an empty JSON array and only accepts valueless assignments. I added a code comment in `POConverters` to make this explicit.

File: scripts/mysql/schema-2.0.0-mysql.sql (hunk 4 above)
Comment (yuqi1129) [reply scripts/mysql/schema-2.0.0-mysql.sql:307]: I'm afraid that it cannot guarantee uniqueness when two threads are inserting the same record at the same time. Is it acceptable?

Comment (roryqi) [reply scripts/mysql/schema-2.0.0-mysql.sql:307]: You’re right. We should keep the database-level uniqueness guarantee and include `tag_value` in the unique key. Since a nullable column in a unique key would still allow duplicate valueless assignments, I’ll use a non-null storage representation for the no-value case and define the unique key on `(tag_id, metadata_object_id, metadata_object_type, tag_value, deleted_at)`.

File: scripts/postgresql/upgrade-1.3.0-to-2.0.0-postgresql.sql [hunk 6]
```diff
@@ -31,3 +31,13 @@ CREATE UNIQUE INDEX IF NOT EXISTS uk_mid_geid_del ON group_meta (metalake_id, ex
...
 ALTER TABLE table_column_version_info
     ALTER COLUMN column_comment TYPE VARCHAR(4096);
+
+ALTER TABLE tag_meta ADD COLUMN IF NOT EXISTS allowed_values TEXT DEFAULT NULL;
+COMMENT ON COLUMN tag_meta.allowed_values IS 'tag allowed values as a JSON string array, NULL allows any value, [] allows no value';
+
+ALTER TABLE tag_relation_meta ADD COLUMN IF NOT EXISTS tag_value VARCHAR(256) DEFAULT NULL;
+COMMENT ON COLUMN tag_relation_meta.tag_value IS 'tag assignment value';
+
+ALTER TABLE tag_relation_meta DROP CONSTRAINT IF EXISTS tag_relation_meta_tag_id_metadata_object_id_metadata_object_key;
```
[the remark this answers is not in the document]
Comment (roryqi) [reply scripts/postgresql/upgrade-1.3.0-to-2.0.0-postgresql.sql:41]: The constraint name in the script is the actual PostgreSQL-generated name. PostgreSQL limits identifiers to 63 bytes and truncates the generated name for `UNIQUE (tag_id, metadata_object_id, metadata_object_type, deleted_at)` to `tag_relation_meta_tag_id_metadata_object_id_metadata_object_key`. Therefore the existing `DROP CONSTRAINT` targets the v1.3.0 constraint; no `DO` block is needed.

File: scripts/postgresql/schema-2.0.0-postgresql.sql [hunk 7]
```diff
@@ -523,21 +525,23 @@ CREATE TABLE IF NOT EXISTS tag_relation_meta (
...
     metadata_object_id BIGINT NOT NULL,
     metadata_object_type VARCHAR(64) NOT NULL,
+    tag_value VARCHAR(256) DEFAULT NULL,
     audit_info TEXT NOT NULL,
     current_version INT NOT NULL DEFAULT 1,
     last_version INT NOT NULL DEFAULT 1,
     deleted_at BIGINT NOT NULL DEFAULT 0,
-    PRIMARY KEY (id),
-    UNIQUE (tag_id, metadata_object_id, metadata_object_type, deleted_at)
+    PRIMARY KEY (id)
```
[the remark this answers is not in the document]
Comment (roryqi) [reply scripts/postgresql/schema-2.0.0-postgresql.sql]: This will be addressed by restoring database-level uniqueness across all three backends. The no-value case also needs a non-null storage representation because nullable columns do not reliably enforce uniqueness. The replacement key will cover `(tag_id, metadata_object_id, metadata_object_type, tag_value, deleted_at)` in both schema and upgrade scripts.

File: core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java [hunk 8]
```diff
@@ -394,6 +514,104 @@ public int deleteTagMetasByLegacyTimeline(long legacyTimeline, int limit) {
...
+  }
+
+  private static TagMetadataObjectRelPO tagRelForValue(
+      TagPO tagPO, Long metadataObjectId, MetadataObject metadataObject, TagValue tagValue) {
+    return POConverters.initializeTagMetadataObjectRelPOWithVersion(
+        tagPO.getTagId(),
+        metadataObjectId,
+        metadataObject.type().toString(),
+        tagValue.value().orElse(null));
+  }
```
[the remark this answers is not in the document]
Comment (roryqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/service/TagMetaService.java:586]: The delete mapper accepts `TagMetadataObjectRelPO` and its SQL reads only `tagId`, `metadataObjectId`, `metadataObjectType`, and `tagValue`; the initialized audit/version fields are not used by the delete statement. Introducing a dedicated one-use parameter type would expand the mapper/provider surface without changing behavior, so I prefer to keep the existing PO construction here.

File: scripts/postgresql/schema-2.0.0-postgresql.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuqi1129) [reply scripts/postgresql/schema-2.0.0-postgresql.sql]: I'm also concerned about this point, and I cannot see how you will handle such scenarios.

File: scripts/mysql/schema-2.0.0-mysql.sql [hunk 9]
```diff
@@ -299,14 +300,16 @@ CREATE TABLE IF NOT EXISTS `tag_relation_meta` (
...
     `metadata_object_id` BIGINT(20) UNSIGNED NOT NULL COMMENT 'metadata object id',
     `metadata_object_type` VARCHAR(64) NOT NULL COMMENT 'metadata object type',
+    `tag_value` VARCHAR(256) NOT NULL DEFAULT '' COMMENT 'tag assignment value, empty string means no value',
     `audit_info` MEDIUMTEXT NOT NULL COMMENT 'tag relation audit info',
     `current_version` INT UNSIGNED NOT NULL DEFAULT 1 COMMENT 'tag relation current version',
     `last_version` INT UNSIGNED NOT NULL DEFAULT 1 COMMENT 'tag relation last version',
     `deleted_at` BIGINT(20) UNSIGNED NOT NULL DEFAULT 0 COMMENT 'tag relation deleted at',
     PRIMARY KEY (`id`),
-    UNIQUE KEY `uk_ti_mi_mo_del` (`tag_id`, `metadata_object_id`, `metadata_object_type`, `deleted_at`),
+    UNIQUE KEY `uk_ti_mi_mo_tv_del` (`tag_id`, `metadata_object_id`, `metadata_object_type`, `tag_value`, `deleted_at`),
```
Comment (yuqi1129) [inline scripts/mysql/schema-2.0.0-mysql.sql:309]: `tag_value` can be null, if two rows are the same and `tag_value` is null, they can still be pushed into the table.

File: scripts/postgresql/schema-2.0.0-postgresql.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (roryqi) [reply scripts/postgresql/schema-2.0.0-postgresql.sql]: Implemented in f71245cff. `tag_value` is now `NOT NULL DEFAULT ''`, and valueless assignments are normalized to the empty string before persistence. The database uniqueness key is `(tag_id, metadata_object_id, metadata_object_type, tag_value, deleted_at)` in the MySQL, H2, and PostgreSQL schemas and upgrade scripts. I also added a database-level regression test that attempts to insert a duplicate active valueless relation and verifies that the constraint rejects it; the relevant tests pass on all three database backends.

File: scripts/mysql/schema-2.0.0-mysql.sql (hunk 9 above)
Comment (roryqi) [reply scripts/mysql/schema-2.0.0-mysql.sql:309]: The updated definition is `tag_value VARCHAR(256) NOT NULL DEFAULT ''`, so `NULL` cannot be stored. The persistence converter also normalizes the API no-value case to `''`. Therefore two otherwise identical valueless rows have the same non-null `tag_value` and are rejected by `uk_ti_mi_mo_tv_del`.

Comment (yuqi1129) [reply scripts/mysql/schema-2.0.0-mysql.sql:309]: OK

**Review by yuqi1129 — APPROVED** [review]
